### PR TITLE
Fix a bug in [caml_modify_generational_global_root]

### PR DESCRIPTION
**Pull request by jhjourdan**
When the old value was static (i.e., boxed, but outside of the major and the minor heaps) but the new value pointed to one of the two heaps, then the root was not re-inserted in the lists.

### CODE REVIEW

**Comment by jhjourdan** [comment]
This is a safety-critical bug for anyone using this API. This should be back-ported to any OCaml version which is still maintained.

**Comment by jhjourdan** [comment]
I pushed a new version of the patch. I makes explicit in a comment the invariant of generational roots, and fixes the dual bug, where a root becomes a pointer outside of the heap, and the root was not removed.

**Comment by dra27** [comment]
The test passes in bytecode - the bug should be triggerable in bytecode too?

**Comment by gasche** [comment]
If I understand correctly, this is a sort of follow-up on your earlier fix to the removal logic for generational roots, #607. (I guess both fixes come from statmemprof hacking.)

**Comment by gasche** [comment]
I looked at the code. It seems to me that:
1. The logic is fairly complex (it was simple at the start, and became complex when fixing bug #4704 which added the initial handling of changes between boxed and unboxed, now tuned by this PR to take out-of-heap blocks into account). 
2. The code is not very efficient due to the systematic uses of the `Is_in_*` macros that access the page table (the new, more correct code makes this worse, by adding a `Is_in*` check after a lot of `Is_block` conditionals).

For example, if you replace a young value by a static value, or a static value by a young value, in both cases, if I read the code correctly, there will be 2 heap-classification checks for the young value, and 2 heap-classification checks for the old value.

I think that both problems could be solved at once by adding a classification function (in the spirit of `Classify_addr(a)`, but tailored to generational roots) that takes a value and returns a value representing whether the address should be untracked, tracked in the young list or tracked in the old list

The resulting code would look as follows (written directly in this message box: unparsed, untested, etc), where it is easy to check that all cases are covered, and at most one address-classification computation is made for each value.

```c
enum gc_root_class {
  YOUNG,
  OLD,
  UNTRACKED
};

enum gc_root_class classify_gc_root(value v);

CAMLexport void caml_modify_global_roots(value *r, value new_val) {
  value old_val = *r;
  enum gc_root_class old_class = classify_gc_root(oldval);
  enum gc_root_class new_class = classify_gc_root(newval);
  if (old_class == YOUNG) {
    // if new_class is OLD, we leave the fixup work to the next minor GC
    if (new_class == UNTRACKED)
      caml_delete_global_root(&caml_global_roots_young, r);
  }
  else if (old_class == OLD) {
    if (new_class == YOUNG) {
     caml_delete_global_root(&caml_global_roots_old, r);
     caml_insert_global_root(&caml_global_roots_young, r);
     // Note: even if old_class is OLD, it may still be in the young roots.
     // The code is still correct in that case, as
     // deleting an absent root and inserting a present root are no-ops.
    } else if (new_class == UNTRACKED) {
      // the old value may still be in the young roots
      caml_delete_global_root(&caml_global_roots_young, r);
      caml_delete_global_root(&caml_global_roots_old, r);
    }
  }
  else { // (old_class == UNTRACKED)
    if (new_class == YOUNG)
      caml_insert_global_root(&caml_global_roots_young, r);
    else if (new_class == OLD)
      caml_insert_global_root(&caml_global_roots_old, r);
  }
  *r = new_val;
}
```

**Comment by jhjourdan** [comment]
> If I understand correctly, this is a sort of follow-up on your earlier fix to the removal logic for generational roots, #607. (I guess both fixes come from statmemprof hacking.)

Interesting, I completely forgot about that one! Indeed, this is the result of statmemprof hacking. Actually, I would say this is a different bug, since that one involves static data, while #607 only involves data in the heap. Hopefully, after 3 fixes, this function is now correct.

**Comment by jhjourdan** [comment]
> The test passes in bytecode - the bug should be triggerable in bytecode too?

The way  the test works in native mode is that `(1,1)` is pre-allocated in static region. But the bytecode compiler does not perform this optimization. I will try with atoms.

**Comment by jhjourdan** [comment]
> I think that both problems could be solved at once by adding a classification function (in the spirit of Classify_addr(a), but tailored to generational roots) that takes a value and returns a value representing whether the address should be untracked, tracked in the young list or tracked in the old list.

Ok, let me try to propose something like this.

**Comment by jhjourdan** [comment]
@gasche, here is a new version, which is hopefully easier to read.

Still, the optimization that lets us not do anything in the young-to-old case makes things significantly more complex. Do we really need it? Is this a use case which is particularly frequent?

File: runtime/globroots.c
Comment (gasche) [inline runtime/globroots.c:221]: It's a bit more low-level, but some parts of the GC do this: you could call `Classify_addr(v)` directly here (see `caml/address_class.h`).

File: runtime/globroots.c
Comment (gasche) [inline runtime/globroots.c:250]: This fallthrough makes semantic sense because `OLD` can be understood as "OLD or maybe still YOUNG". (Nothing to change here.)

File: runtime/globroots.c
Comment (gasche) [inline runtime/globroots.c]: Contrary to the one above, I'm not fond of this fallthrough here which is purely for code factorization, but does not make much sense to me. I would recommend to either duplicate the `caml_insert` call -- no problem with that, or have a structure of the form

```c
if (OLD) ...;
if !(YOUNG) ...;
```

which is more explicit about why things are shared. (But that requires declaring an intermediate variable, so it is different from your style in this function.)

File: runtime/globroots.c
Comment (gasche) [inline runtime/globroots.c:283]: Nice sharing!

**Review by gasche — COMMENTED** [review]
Yes, I think that the new implementation improves things (`switch` in C is clearly the semantically-appropriate construct, but yet never really pleasant to use...). See comments inline.

File: runtime/globroots.c
Comment (gasche) [reply runtime/globroots.c]: Actually I think just using `default:` instead of `case UNTRACKED:` would be clear enough (with the fallthrough); `default` means `non-YOUNG` here.

File: runtime/globroots.c
Comment (jhjourdan) [reply runtime/globroots.c:221]: I still prefer this version, for two reasons:
- As you said, `Classify_addr` is lower level, and not part of the public interface of `caml/address_class.h`
- This version avoids a page table lookup when the value is unboxed or young.

File: runtime/globroots.c
Comment (jhjourdan) [reply runtime/globroots.c]: New version pushed. I still prefer the `if` version. Your intuition for `default` is rather weird.

File: runtime/globroots.c
Comment (gasche) [reply runtime/globroots.c:221]: Indeed, I confused the lookup semantics of `Is_young` and `Is_in_heap_or_young`. You are correct that this version is better then.

File: runtime/globroots.c
Comment (gasche) [reply runtime/globroots.c]: Yes, the new version is nice! Thanks.

**Comment by gasche** [comment]
I like the new implementation, thanks! I'd like to give a full re-read of the code before approving, and probably won't have time before tomorrow.

**Review by gasche — APPROVED** [review]
I re-read the PR and I am convinced that the new code is correct.

On the question of whether the complexity of tolerating old values in the young list is worth it: I am, as @jhjourdan, unconvinced.

This optimization speeds up modifications of roots in the case where we are replacing a young value with an old value, at the cost of slowing down deletion of old values, which must also be removed from the young list.

I don't have a clear mental model of the life cycle we can expect for generational roots, but I would expect that the use-case is to write (at registration time and a root-modification time) untracked constants and young values that later become old, rather than old values, and that deletion is more common than modification. So this optimization is not a clear win.

I looked for usages of the generational API online, and I found:

- Some usages to register OCaml-shaped data created from the C side, such as [elm_box_wrap.c in ocaml-efl](https://github.com/axiles/ocaml-efl/blob/04acd12a6ad78392740efbd626867091825f7f74/src/elm_box_wrap.c#L49), where the registered values are young. (I haven't found any instance where the roots are un-registered, so there may be a leak in this code.)

- Some usages to register callbacks for various C APIs, [ide_mac_stubs.c in CoqIDE](https://github.com/mattam82/Coq-misc/blob/60bc3cbe72083f4fa1aa759914936e4fa3d6b42e/ide/ide_mac_stubs.c), [gnutls.c in ocamlnet](https://github.com/UnixJunkie/ocamlnet/blob/81282279d073a85e3ed6e8b7030710ed9b7b2d02/code/src/nettls-gnutls/gnutls.c#L466-L474): in this case I expect the closure to be constructed from the OCaml side as an argument or defined close by, so either static (if closed) or young (if just allocated)

- Generic usages in C++ constructor/destructor pairs in generic interoperability libraries like [this usage in mlsmoke](https://github.com/pippijn/mlsmoke/blob/feabfa195c491603a0e7078b6198ada744bb2957/mlsmoke/ml_qlist.cpp). In this case it's hard to predict the class of the registered value, but there are only deletions, no modifications.

That said, it seems that making this optimization is outside the scope of the present PR, so I would suggest that we deal with the bugfix and code clarification now, and that you write a second PR to remove the optimization -- if you care -- on top of this one, or after this one is merged.

**Comment by jhjourdan** [comment]
> That said, it seems that making this optimization is outside the scope of the present PR, so I would suggest that we deal with the bugfix and code clarification now, and that you write a second PR to remove the optimization -- if you care -- on top of this one, or after this one is merged.

LGTM. Not sure I do care, though.

**Comment by gasche** [comment]
I just started a `precheck` test (I was mildly annoyed by the unrelated Appveyor failure: there is a small chance that the new code fails with some C compiler somewhere), and will merge if it passes.

https://ci.inria.fr/ocaml/job/precheck/259/

**Comment by jhjourdan** [comment]
Thanks!

Don't you think this should be backported to e.g., the 4.09 branch?

**Comment by gasche** [comment]
Yes for 4.09 (I cherry-picked in 2e9e3264cf66cff16e00cad1df55aad4c3f55280), no for 4.08.

**Comment by jhjourdan** [comment]
@gasche : Could you add yourself as a reviewer in the `Changes` file? Currently, it shows `???`.

**Pull request merged**
